src/providers/helpers/saveLocalObservation.js

```diff
@@ -0,0 +1,32 @@
+// @flow
+
+import Realm from "realm";
+
+import realmConfig from "../../models/index";
+import Observation from "../../models/Observation";
+import { getUserId } from "../../components/LoginSignUp/AuthenticationService";
+
+const saveLocalObservation = async ( currentObs: Object ): Promise<any> => {
+  try {
+    const realm = await Realm.open( realmConfig );
+
+    // make sure local observations have user details for ObsDetail
+    const id = await getUserId( );
+    const user = realm.objectForPrimaryKey( "User", Number( id ) );
+    currentObs.user = user;
+
+    const obsToSave = Observation.saveLocalObservationForUpload( currentObs, realm );
+    realm?.write( ( ) => {
+      // using 'modified' here for the case where a new observation has the same Taxon
+      // as a previous observation; otherwise, realm will error out
+      realm?.create( "Observation", obsToSave, "modified" );
+    } );
+    const newLocalObs = realm.objectForPrimaryKey( "Observation", currentObs.uuid );
+    return newLocalObs;
+  } catch ( e ) {
+    console.log( e, "couldn't save observation locally" );
+    return null;
+  }
+};
+
+export default saveLocalObservation;
```

Comment:
@albullington you said

> I didn't move 100% of the saveLocalObservation code to the Observation model, because doing so results in a require-cycle error where realm is trying to open the realmConfig schema from index.js, while index.js is also requiring the Observation model.

I think there's a way to deal with this but I need to think about it, so let's consider this non-blocking.

Reply:
I still don't think this needs to block, but here are the results of some further thinking:

* The [`realm-context` branch](https://github.com/inaturalist/iNaturalistReactNative/compare/save-observations...realm-context) shows how we could use a context to provide a single Realm instance to the entire app. This gets around the require cycle error and centralizes the opening and closing of Realm, which seems a bit DRYer to me, but it also involves passing this realm instance to the models or to helpers every time they need to read / write data, which feels very awkward.
* The [`realm-in-models` branch](https://github.com/inaturalist/iNaturalistReactNative/compare/realm-context...realm-in-models) is even messier, but it extends `realm-context` by storing a reference to realm as a class variable of a superclass to all models, allowing them all to interact with Realm and to abstract Realm away from the components (which may or may not be good, but it certainly allows the models to perform their duty as gatekeepers between the application and the data). This is very loosely based on ActiveRecord in Rails, and most of what's interesting is in the `Record` and `Observation` models.

While doing this, I also realized that [Realm actually provides its own `RealmProvider`](https://www.mongodb.com/docs/realm/sdk/react-native/use-realm-react/#using-the-realm-provider) and some associated helpers that we might want to use. The former is basically the same as my `RealmProvider`, I think.

I'm also still pretty suspicious of the `ObsEditContext` and practices like adding a new observation to the context before navigating to `ObsEdit` which then accesses that obs from the context. In my mind, navigation should just navigate, and `ObsEdit` should be responsible for making. new observation or loading an existing one. Again, doesn't need to block this PR, but we should think about separation of concerns after we merge.